### PR TITLE
migration contract

### PR DESCRIPTION
**Pull request by reganwrench**

### CODE REVIEW

File: contracts/veMigration.sol [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        for (uint256 i = 0; i < feeDistributors.length; i++) {
+            IFeeDistributor(feeDistributors[i]).claim_many(tokenIds);
+        }
+        MigrationLock[] memory migrationLocks = new MigrationLock[](tokenIds.length);
+        uint256 oneWeekFromNow = block.timestamp + WEEK;
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            require(IVotingEscrow(veIB).ownerOf(tokenIds[i]) == msg.sender, "You are not the owner of this token");
+            IVotingEscrow.LockedBalance memory lockBalance = IVotingEscrow(veIB).locked(tokenIds[i]);
+            uint256 duration = lockBalance.end >= oneWeekFromNow ? lockBalance.end - block.timestamp : WEEK;
+            migrationLocks[i] = MigrationLock(uint256(uint128(lockBalance.amount)), duration);
```
Comment (ibsunhub) [inline contracts/veMigration.sol]: Why wrap to uint128 and then wrap to uint256 later?

File: contracts/veMigration.sol [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        require(_sender != address(0), "_sender address cannot be 0");
+        require(_receiver != address(0), "_receiver address cannot be 0");
+        sender = _sender;
+        receiver = _receiver;
+    }
+
+    /// @notice function to initiate migration on source chain, it help users claim rewards from fee_distibutors
+    ///         and then burn the veIB NFTs before initiating the anyCall to destination chain
+    /// @param tokenIds array of tokenIds to migrate
+    function migrate(uint256[] calldata tokenIds) external {
```
Comment (ibsunhub) [inline contracts/veMigration.sol]: Could add reentry protection on this function.

File: contracts/veMigration.sol [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+            // execute migration flow on destination chain
+            executeMigration(data);
+        }
+        return (true, "");
+    }
+
+    /// @notice function to log failure on source chain, when the migration call on the destination chain is unsuccessful
+    /// @param _initialCallTo initial call to address on the destination chain
+    /// @param _initialCallData initial calldata sent to the destination chain
+    function anyFallback(address _initialCallTo, bytes calldata _initialCallData) external onlySelf {
```
Comment (ibsunhub) [inline contracts/veMigration.sol]: Change this function's visibility to internal?

**Review by ib-tycho — APPROVED** [review]

File: hardhat.config.ts [hunk 4]
```diff
@@ -52,11 +51,19 @@ const config: HardhatUserConfig = {
     hardhat: hardhatNetworkConfig,
     optimism: {
       url: `https://optimism-mainnet.infura.io/v3/${process.env.INFURA_API_KEY}`,
-      accounts: [`${process.env.OP_PRIVATE_KEY}`],
+      accounts: [`${process.env.DEPLOYER_PRIVATE_KEY}`],
```
Comment (ib-tycho) [inline hardhat.config.ts]: handle invalid env

accounts: [`0x${process.env.DEPLOY_PRIVATE_KEY ?? ''}`]

File: README.md [hunk 5]
```diff
@@ -25,10 +25,22 @@ npm run coverage
 ```
 
 ## Deployment
-```
-npm run deployOP
-npm run deployFTM
-```
+1. In ```000_deploy_dest_chain.ts```, fill in the ```veIB``` address
```
Comment (ib-tycho) [inline README.md]: should put veIB as `namedAccounts` in `hardhat.config.ts`

example: https://github.com/ibdotxyz/ib-deployment/blob/op/hardhat.config.ts

File: README.md [hunk 6]
```diff
@@ -25,10 +25,22 @@ npm run coverage
...
+2. Execute:
+    ```bash
+    npx hardhat deploy --tags deploy-dest --network optimism
+    ```
+3. In ```001_deploy_src_chain.ts```, fill in the ```receiver``` address (veMigrationDest deployed in Step 2)
```
Comment (ib-tycho) [inline README.md:33]: get deployed contract address using `get`: https://github.com/ibdotxyz/ib-deployment/blob/op/deploy/007_set_comptroller.ts#L11

might need to check out how to get from different network

File: hardhat.config.ts (hunk 4 above)
Comment (reganwrench) [reply hardhat.config.ts]: done

File: README.md (hunk 5 above)
Comment (reganwrench) [reply README.md]: done

File: contracts/veMigrationDest.sol [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    /// @param oldTokenIds old tokenIds of the user
+    /// @param newTokenIds new tokenIds of the user after migration
+    event MigrationCompleted(address user, uint256[] oldTokenIds, uint256[] newTokenIds);
+
+    modifier onlyExecutor() {
+        require(msg.sender == anycallExecutor, "Only executor can call this function");
+        _;
+    }
+
+    modifier onlySelf() {
```
Comment (ibsunhub) [inline contracts/veMigrationDest.sol]: unused

**Review by ibsunhub — APPROVED** [review]

File: contracts/veMigrationDest.sol [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+contract veMigrationDest is Ownable, ReentrancyGuard {
+    address public immutable ibToken;
+    address public immutable anycallExecutor;
+    address public immutable anyCall;
+    address public immutable veIB;
+    uint256 public immutable srcChainId;
+
+    address public sender;
+
+    uint256 public constant WEEK = 1 weeks;
```
Comment (ibsunhub) [inline contracts/veMigrationDest.sol]: unused

File: contracts/veMigrationDest.sol [hunk 9]
```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+import "./interfaces/IVotingEscrow.sol";
+import "./interfaces/IFeeDistributor.sol";
```
Comment (ibsunhub) [inline contracts/veMigrationDest.sol]: unused import

File: contracts/veMigrationSrc.sol [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+import "./interfaces/IExecutor.sol";
+import "./interfaces/IERC20.sol";
+
+struct MigrationLock {
+    uint256 amount;
+    uint256 duration;
+}
+
+contract veMigrationSrc is Ownable {
+    address public immutable ibToken;
```
Comment (ibsunhub) [inline contracts/veMigrationSrc.sol]: ditto, unused

Comment (reganwrench) [reply contracts/veMigrationSrc.sol]: fixed

File: contracts/veMigrationDest.sol (hunk 9 above)
Comment (reganwrench) [reply contracts/veMigrationDest.sol]: fixed

File: contracts/veMigrationDest.sol (hunk 8 above)
Comment (reganwrench) [reply contracts/veMigrationDest.sol]: fixed

File: contracts/veMigrationDest.sol (hunk 7 above)
Comment (reganwrench) [reply contracts/veMigrationDest.sol]: fixed

**Review by ibsunhub — APPROVED** [review]

File: deploy/001_deploy_src_chain.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  const { deployer, ibToken, veIB, anyCall } = await getNamedAccounts();
+
+  // TODO: fill in deployed contract address on op
+  const srcChainId = 4; // ftm 250 ftm testNet 4002
+  const destChainId = 4002; // optimism 10, rinkeby 4
+  // const srcChainId = 250;
+  // const destChainId = 10;
+  // usdcFeeDist, solidFeeDist, IBFeeDist
+  // const feeDistributors = ["0x31A1D83C715F4bd6fE7A26f1Ce279Cec15011AE9", "0xB634c662296a4BA117A422bFE6742B75989Bd714", "0x3Af7c11d112C1C730E5ceE339Ca5B48F9309aCbC"];
+  const feeDistributors = [];
```
Comment (ib-tycho) [inline deploy/001_deploy_src_chain.ts]: change to production ready script

Comment (reganwrench) [reply deploy/001_deploy_src_chain.ts]: done

File: hardhat.config.ts [hunk 12]
```diff
@@ -28,12 +28,29 @@ const hardhatNetworkConfig =
...
     },
+    ibToken: {
+      fantom: "0x00a35FD824c717879BF370E70AC6868b95870Dfb",
+      optimism: "0x00a35FD824c717879BF370E70AC6868b95870Dfb",
+      ftmTest: "0x534F54Ac33bA40fbA923309C22e2F9b32BF4747a",
+      rinkeby: "0x779EeC7212aa2E35942B1422e12b44510E740F94",
+    },
+    veIB: {
+      fantom: "0xBe33aD085e4a5559e964FA8790ceB83905062065",
+      optimism: "",
```
Comment (ib-tycho) [inline hardhat.config.ts]:
```suggestion
      optimism: "0x707648dfbf9df6b0898f78edf191b85e327e0e05",
```

**Pull request merged**